File: apps/system/api/domain/workspace/repository.go
//go:generate gomockhandler -source=$GOFILE -destination=mock_$GOFILE -package=$GOPACKAGE
package workspace

import (
	"context"

	"github.com/ryo034/react-go-template/apps/system/api/domain/shared/account"
	"github.com/ryo034/react-go-template/apps/system/api/domain/workspace/invitation"
	"github.com/ryo034/react-go-template/apps/system/api/domain/workspace/member"
	"github.com/uptrace/bun"
)

type Repository interface {
	FindAll(ctx context.Context, exec bun.IDB, aID account.ID) (Workspaces, error)
	Create(ctx context.Context, exec bun.IDB, w *Workspace) (*Workspace, error)
	Update(ctx context.Context, exec bun.IDB, w *Workspace) error
	AddMember(ctx context.Context, exec bun.IDB, w *Workspace, m *member.Member) (*member.Member, error)
	UpdateMemberRole(ctx context.Context, exec bun.IDB, assignor *member.Member, m *member.Member) (*member.Member, error)
	FindMember(ctx context.Context, exec bun.IDB, memID member.ID) (*member.Member, error)
	FindAllMembers(ctx context.Context, exec bun.IDB, wID ID) (member.Members, error)
	InviteMembers(ctx context.Context, exec bun.IDB, inviter Inviter, is invitation.Invitations) error
	FindInviterFromToken(ctx context.Context, exec bun.IDB, token invitation.Token) (Inviter, error)
	FindActiveInvitation(ctx context.Context, exec bun.IDB, id invitation.ID) (*invitation.Invitation, *Workspace, error)
	FindAllInvitations(ctx context.Context, exec bun.IDB, wID ID) (invitation.Invitations, error)
	Leave(ctx context.Context, exec bun.IDB, executorID member.ID, mID member.ID) error
}

File: apps/system/client/src/domain/workspace/repository.ts
import {
  Invitation,
  Invitations,
  Invitees,
  Member,
  MemberId,
  Members,
  SelectableRole,
  Workspace,
  WorkspaceId,
  WorkspaceName,
  WorkspaceSubdomain
} from "~/domain"
import { PromiseResult } from "~/infrastructure/shared"

export interface WorkspaceCreateInput {
  subdomain: WorkspaceSubdomain
}

export interface WorkspaceRepository {
  create(i: WorkspaceCreateInput): PromiseResult<Workspace, Error>
  findAllMembers(): PromiseResult<Members, Error>
  inviteMembers(invitees: Invitees): PromiseResult<Invitations, Error>
  findAllInvitations(): PromiseResult<Invitations, Error>
  resendInvitation(invitation: Invitation): PromiseResult<null, Error>
  revokeInvitation(invitation: Invitation): PromiseResult<Invitations, Error>
  updateMemberRole(memberId: MemberId, role: SelectableRole): PromiseResult<Member, Error>
  updateWorkspace(
    workspaceId: WorkspaceId,
    name: WorkspaceName,
    subdomain: WorkspaceSubdomain
  ): PromiseResult<Workspace, Error>
}
